### PR TITLE
[Master] Add Internal Token Revocation implementation for token persistence removal feature

### PR DESCRIPTION
**Pull request by Lakith-Rambukkanage**
## Purpose
This PR adds the implementation for the indirect token revocations due to internal events. The internal events are of two types.
- Application Events (Consumer Key Events) [1]
- User Events [2]

With the token persistence removal feature [3], such revocations has to be handled in the form of attribute(rule) base validation, since the tokens are not available in the db. Therefore the attributes to be considered for the revocation is propagated to CP from the KM then to the GW cache.

[1] - https://github.com/wso2/api-manager/issues/1678
[2] - https://github.com/wso2/api-manager/issues/2159
[3] - https://github.com/wso2/api-manager/issues/1664

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/jwt/JWTValidator.java [hunk 1]
```diff
@@ -176,6 +183,42 @@ public AuthenticationContext authenticate(SignedJWTInfo signedJWTInfo, MessageCo
...
+                if (log.isDebugEnabled()) {
+                    log.debug("Consumer key retrieved from the  jwt token map is in revoked consumer key map." +
+                            " Token: " + GatewayUtils.getMaskedToken(jwtHeader));
+                }
+                log.error("Invalid JWT token. " + GatewayUtils.getMaskedToken(jwtHeader));
+                throw new APISecurityException(APISecurityConstants.API_AUTH_INVALID_CREDENTIALS,
+                        "Invalid JWT token");
+            }
+            if (jwtGeneratedTime != 0 && signedJWTInfo.getJwtClaimsSet().getSubject()
+                    .equals(signedJWTInfo.getJwtClaimsSet().getClaim("client_id"))
```
Comment (tharikaGitHub) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/jwt/JWTValidator.java:198]: Better to use a string constant for "client_id" here.

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/jwt/InternalRevokedJWTDataHolder.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
+/*
+ *  Copyright (c) 2023, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (tharikaGitHub) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/jwt/InternalRevokedJWTDataHolder.java:2]:
```suggestion
 *  Copyright (c) 2023, WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
```

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/jwt/RevokeConditionsRetriever.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
+/*
+ *  Copyright (c) 2019, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (tharikaGitHub) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/jwt/RevokeConditionsRetriever.java:2]:
```suggestion
 *  Copyright (c) 2023, WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
```

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/listeners/GatewayStartupListener.java [hunk 4]
```diff
@@ -31,7 +31,7 @@
 import org.wso2.carbon.apimgt.gateway.InMemoryAPIDeployer;
 import org.wso2.carbon.apimgt.gateway.internal.DataHolder;
 import org.wso2.carbon.apimgt.gateway.internal.ServiceReferenceHolder;
-import org.wso2.carbon.apimgt.gateway.jwt.RevokedJWTTokensRetriever;
+import org.wso2.carbon.apimgt.gateway.jwt.*;
```
Comment (tharikaGitHub) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/listeners/GatewayStartupListener.java:34]: Let's remove wild card imports

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/listeners/GatewayTokenRevocationMessageListener.java [hunk 5]
```diff
@@ -47,16 +48,30 @@ public void onMessage(Message message) {
...
                              * revokedToken - Revoked Token which should be removed from the cache
                              * expiryTime - ExpiryTime of the token if token is JWT, otherwise expiry is set to 0
                              */
-                            handleRevokedTokenMessage(payloadData.get(APIConstants.REVOKED_TOKEN_KEY).asText(),
+                                handleRevokedTokenMessage(payloadData.get(APIConstants.REVOKED_TOKEN_KEY).asText(),
                                     payloadData.get(APIConstants.REVOKED_TOKEN_EXPIRY_TIME).asLong(),
                                     payloadData.get(APIConstants.REVOKED_TOKEN_TYPE).asText());
                         }
+
+                        if (payloadData.get("type") != null && payloadData.get("type").asText()
```
Comment (tharikaGitHub) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/listeners/GatewayTokenRevocationMessageListener.java:63]: Let's use string constants where possible.

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/listeners/ServerStartupListener.java [hunk 6]
```diff
@@ -21,9 +21,8 @@
 import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
 import org.wso2.carbon.apimgt.gateway.internal.ServiceReferenceHolder;
-import org.wso2.carbon.apimgt.gateway.service.APIThrottleDataServiceImpl;
+import org.wso2.carbon.apimgt.gateway.service.*;
```
Comment (tharikaGitHub) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/listeners/ServerStartupListener.java:24]: Remove wild card imports

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java [hunk 7]
```diff
@@ -143,21 +143,7 @@
...
-import java.util.HashMap;
-import java.util.HashSet;
-import java.util.LinkedHashMap;
-import java.util.LinkedHashSet;
-import java.util.List;
-import java.util.Map;
-import java.util.Set;
-import java.util.TreeMap;
-import java.util.UUID;
+import java.util.*;
```
Comment (tharikaGitHub) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:146]: Wild card imports

File: components/apimgt/org.wso2.carbon.apimgt.notification/src/main/java/org/wso2/carbon/apimgt/notification/event/InternalTokenRevocationUserEvent.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
+package org.wso2.carbon.apimgt.notification.event;
```
Comment (tharikaGitHub) [inline components/apimgt/org.wso2.carbon.apimgt.notification/src/main/java/org/wso2/carbon/apimgt/notification/event/InternalTokenRevocationUserEvent.java:1]: License header

File: features/apimgt/org.wso2.carbon.apimgt.core.feature/src/main/resources/sql/h2.sql [hunk 9]
```diff
@@ -2361,8 +2361,20 @@ CREATE TABLE IF NOT EXISTS AM_CORRELATION_PROPERTIES(
     FOREIGN KEY (COMPONENT_NAME) REFERENCES AM_CORRELATION_CONFIGS(COMPONENT_NAME) ON DELETE CASCADE
 );
 
+  CREATE TABLE IF NOT EXISTS AM_INTERNAL_TOKEN_REVOCATION_CONSUMER_KEY_EVENTS (
```
Comment (tharikaGitHub) [inline features/apimgt/org.wso2.carbon.apimgt.core.feature/src/main/resources/sql/h2.sql:2364]: We will need to add these tables to other DB scripts as well.

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/jwt/JWTValidator.java [hunk 10]
```diff
@@ -43,7 +43,7 @@
 import org.wso2.carbon.apimgt.gateway.handlers.security.AuthenticationContext;
 import org.wso2.carbon.apimgt.gateway.handlers.streaming.websocket.WebSocketApiConstants;
 import org.wso2.carbon.apimgt.gateway.internal.ServiceReferenceHolder;
-import org.wso2.carbon.apimgt.gateway.jwt.RevokedJWTDataHolder;
+import org.wso2.carbon.apimgt.gateway.jwt.*;
```
[the remark this answers is not in the document]
Comment (Lakith-Rambukkanage) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/jwt/JWTValidator.java]: Fixed by. https://github.com/wso2/carbon-apimgt/pull/12174/commits/70b076a6067da5daa54577ca7065398e9a9c5e7d

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/jwt/JWTValidator.java (hunk 1 above)
Comment (Lakith-Rambukkanage) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/jwt/JWTValidator.java]: Fixed by. https://github.com/wso2/carbon-apimgt/pull/12174/commits/70b076a6067da5daa54577ca7065398e9a9c5e7d

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/jwt/InternalRevokedJWTDataHolder.java (hunk 2 above)
Comment (Lakith-Rambukkanage) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/jwt/InternalRevokedJWTDataHolder.java]: Fixed by. https://github.com/wso2/carbon-apimgt/pull/12174/commits/70b076a6067da5daa54577ca7065398e9a9c5e7d

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/jwt/RevokeConditionsRetriever.java (hunk 3 above)
Comment (Lakith-Rambukkanage) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/jwt/RevokeConditionsRetriever.java]: Fixed by. https://github.com/wso2/carbon-apimgt/pull/12174/commits/70b076a6067da5daa54577ca7065398e9a9c5e7d

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/listeners/GatewayStartupListener.java (hunk 4 above)
Comment (Lakith-Rambukkanage) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/listeners/GatewayStartupListener.java]: Fixed by. https://github.com/wso2/carbon-apimgt/pull/12174/commits/70b076a6067da5daa54577ca7065398e9a9c5e7d

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/listeners/GatewayTokenRevocationMessageListener.java (hunk 5 above)
Comment (Lakith-Rambukkanage) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/listeners/GatewayTokenRevocationMessageListener.java]: Fixed by. https://github.com/wso2/carbon-apimgt/pull/12174/commits/70b076a6067da5daa54577ca7065398e9a9c5e7d

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/listeners/ServerStartupListener.java (hunk 6 above)
Comment (Lakith-Rambukkanage) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/listeners/ServerStartupListener.java]: Fixed by. https://github.com/wso2/carbon-apimgt/pull/12174/commits/70b076a6067da5daa54577ca7065398e9a9c5e7d

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java (hunk 7 above)
Comment (Lakith-Rambukkanage) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java]: Fixed by. https://github.com/wso2/carbon-apimgt/pull/12174/commits/70b076a6067da5daa54577ca7065398e9a9c5e7d

File: components/apimgt/org.wso2.carbon.apimgt.notification/src/main/java/org/wso2/carbon/apimgt/notification/event/InternalTokenRevocationUserEvent.java (hunk 8 above)
Comment (Lakith-Rambukkanage) [reply components/apimgt/org.wso2.carbon.apimgt.notification/src/main/java/org/wso2/carbon/apimgt/notification/event/InternalTokenRevocationUserEvent.java:19]: Fixed by. https://github.com/wso2/carbon-apimgt/pull/12174/commits/70b076a6067da5daa54577ca7065398e9a9c5e7d

File: features/apimgt/org.wso2.carbon.apimgt.core.feature/src/main/resources/sql/h2.sql (hunk 9 above)
Comment (Lakith-Rambukkanage) [reply features/apimgt/org.wso2.carbon.apimgt.core.feature/src/main/resources/sql/h2.sql]: Fixed by. https://github.com/wso2/carbon-apimgt/pull/12174/commits/70b076a6067da5daa54577ca7065398e9a9c5e7d

File: components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/java/org/wso2/carbon/apimgt/internal/service/impl/RevokeConditionsApiServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,18 @@
+package org.wso2.carbon.apimgt.internal.service.impl;
```
[the remark this answers is not in the document]
Comment (Lakith-Rambukkanage) [reply components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/java/org/wso2/carbon/apimgt/internal/service/impl/RevokeConditionsApiServiceImpl.java:19]: Fixed by. https://github.com/wso2/carbon-apimgt/pull/12174/commits/70b076a6067da5daa54577ca7065398e9a9c5e7d

**Pull request closed without merging**